runtime/syntax/javascript.vim

```diff
@@ -40,7 +40,7 @@ syn region  javaScriptStringT	       start=+`+  skip=+\\\\\|\\`+  end=+`+	contai
 syn region  javaScriptEmbed	       start=+${+  end=+}+	contains=@javaScriptEmbededExpr
 
 syn match   javaScriptSpecialCharacter "'\\.'"
-syn match   javaScriptNumber	       "-\=\<\d\+L\=\>\|0[xX][0-9a-fA-F]\+\>"
+syn match   javaScriptNumber	       "-\=\<\d\(_\?\d\)*\>\|0[xX][0-9a-fA-F_]\+\>"
```

Comment:
This is making still the same mistake for hexadecimal numbers where it allows leading, trailing, and duplicated underscores in the middle. It seems that [JavaScript specification](https://github.com/tc39/proposal-numeric-separator/blob/main/spec.md) does not permit for underscores between `0x` and a number (Python does). So based on my implementation for Python, I'd say that the changes should be for the decimal numbers:
```
\d\%(_\=\d\)*
```
Notice the use of `\%( ... \)` for speedup. These are non-capturing _group_ parentheses.
The use of `\=` is the same as `\?` but more typical in Vim scripts.

And for the hexadecimal numbers:
```
0[xX]\x\%(_\=\x\)*
```

Finally, it seems that you dropped optional `L` from the original.
I don't know if that was intentional and does JavaScript still allow `L`, but let's assume it does.
It also looks that the original had an unnecessary close of the word boundary before the `\|` and then another one at the end, but just that one at the end is needed.
So, overall, this expression needs to be, AFAICS,
```
syn match   javaScriptNumber	       "-\=\<\d\%(_\=\d\)*L\=\|0[xX]\x\%(_\=\x\)*\>"
```
If you intentionally dropped `L` there and if JavaScript does not allow it any more (similar to Python 3), then you can remove `L\=` from the expression.